plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,340 @@
+    private final String namespace;
+    private final Set<QueryType> includeQueryTypes;
+    private final ObjectMapper queryStatisticsMapper;
+    private static final Logger logger = Logger.get(OpenLineageListener.class);
+
+    @Inject
+    public OpenLineageListener(OpenLineageClient client, OpenLineageListenerConfig listenerConfig)
+            throws Exception
+    {
+        this.client = client;
+        this.namespace = listenerConfig.getTrinoURI().toString().replace(listenerConfig.getTrinoURI().getScheme(), "trino");
+
+        this.includeQueryTypes = listenerConfig.getIncludeQueryTypes();
+
+        this.queryStatisticsMapper = createQueryStatisticsObjectMapper();
+    }
+
+    @Override
+    public void queryCreated(QueryCreatedEvent queryCreatedEvent)
+    {
+        if (queryTypeSupported(queryCreatedEvent.getContext())) {
+            UUID runID = getQueryId(queryCreatedEvent.getMetadata());
+
+            try {
+                RunEvent event = getStartEvent(runID, queryCreatedEvent);
+                client.emit(event);
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        else {
+            logger.debug(format("Query type %s not supported. Supported query types %s",
+                    queryCreatedEvent.getContext().getQueryType().toString(),
+                    this.includeQueryTypes));
+        }
+    }
+
+    @Override
+    public void queryCompleted(QueryCompletedEvent queryCompletedEvent)
+    {
+        if (queryTypeSupported(queryCompletedEvent.getContext())) {
+            UUID runID = getQueryId(queryCompletedEvent.getMetadata());
+
+            try {
+                RunEvent event = getCompletedEvent(runID, queryCompletedEvent);
+                client.emit(event);
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        else {
+            logger.debug(format("Query type %s not supported. Supported query types %s",
+                    queryCompletedEvent.getContext().getQueryType().toString(),
+                    this.includeQueryTypes));
+        }
+    }
+
+    private Boolean queryTypeSupported(QueryContext queryContext)
+    {
+        if (queryContext.getQueryType().isPresent()) {
+            return this.includeQueryTypes.contains(queryContext.getQueryType().get());
+        }
+        else {
+            return false;
+        }
+    }
+
+    private UUID getQueryId(QueryMetadata queryMetadata)
+    {
+        return UUID.nameUUIDFromBytes(queryMetadata.getQueryId().getBytes(StandardCharsets.UTF_8));
+    }
+
+    private RunFacet getTrinoQueryContextFacet(QueryContext queryContext)
+    {
+        RunFacet queryContextFacet = openLineage.newRunFacet();
+
+        queryContextFacet
+                .getAdditionalProperties()
+                .put("serverAddress", queryContext.getServerAddress());
```

Comment:
Is it possible to extract them to a method so that we could pass each individual argument or pass a map of additional properties. WDYT ?

Reply:
I don't fully get the question, can you rephrase please?

Reply:
how about 

```
    private RunFacet getTrinoQueryContextFacet(QueryContext queryContext)
    {
        RunFacet queryContextFacet = openLineage.newRunFacet();

        java.util.Map<String, String> properties = Maps.<String, String>newHashMap();

        properties.put("server_address", queryContext.getServerAddress());
        properties.put("environment", queryContext.getEnvironment());

        queryContext.getQueryType().ifPresent(queryType ->
                properties.put("query_type", queryType.toString()));

        queryContextFacet
                .getAdditionalProperties()
                .putAll(properties);

        return queryContextFacet;
    }
```

and same for all 3 methods creating facets?